src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/ReadOnlySpanND.cs

```diff
@@ -0,0 +1,468 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Runtime.InteropServices.Marshalling;
+using System.Runtime.Versioning;
+using System.Text;
+using EditorBrowsableAttribute = System.ComponentModel.EditorBrowsableAttribute;
+using EditorBrowsableState = System.ComponentModel.EditorBrowsableState;
+
+#pragma warning disable 0809  //warning CS0809: Obsolete member 'Span<T>.Equals(object)' overrides non-obsolete member 'object.Equals(object)'
+
+namespace System.Numerics.Tensors
+{
+    /// <summary>
+    /// ReadOnlySpanND represents a contiguous region of arbitrary memory. Unlike arrays, it can point to either managed
+    /// or native memory, or to memory allocated on the stack. It is type-safe and memory-safe.
+    /// </summary>
+    [DebuggerTypeProxy(typeof(SpanNDDebugView<>))]
+    [DebuggerDisplay("{ToString(),raw}")]
+    public readonly ref struct ReadOnlySpanND<T>
+    {
+        /// <summary>A byref or a native ptr.</summary>
+        internal readonly ref T _reference;
+        /// <summary>The number of elements this SpanND contains.</summary>
+        internal readonly nint _linearLength;
+        /// <summary>The lengths of each dimension.</summary>
+        internal readonly ReadOnlySpan<nint> _lengths;
+        /// <summary>The strides representing the memory offsets for each dimension.</summary>
+        private readonly ReadOnlySpan<nint> _strides;
+        /// <summary>If the backing memory is permanently pinned (so not just using a fixed statement).</summary>
+        private readonly bool _isPinned = false;
+
+        /// <summary>
+        /// Creates a new read-only span over the entirety of the target array.
+        /// </summary>
+        /// <param name="array">The target array.</param>
+        /// <param name="lengths">The <see cref="System.ReadOnlySpan{T}"/> with the lengths of each dimension.</param>
+        /// <remarks>Returns default when <paramref name="array"/> is null.</remarks>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public ReadOnlySpanND(T[]? array, ReadOnlySpan<nint> lengths)
+        {
+            if (array == null)
+            {
+                this = default;
+                return; // returns default
+            }
+
+            _linearLength = SpanNDHelpers.CalculateTotalLength(lengths);
+            if (_linearLength != array.Length)
+                ThrowHelper.ThrowArgument_LengthsMustEqualArrayLength();
```

Comment:
nit: Argument validation should be done before field assignments, otherwise users who catch the exception may be able to observe the torn state and that could cause bugs.

Reply:
This could be done by simply caching `linearLength` into a local, doing the validation, and then assigning the fields

Reply:
KK. Made the change.